scrapy_poet/utils.py

```diff
@@ -121,27 +121,31 @@ def is_response_going_to_be_used(request, spider):
     if is_callback_using_response(callback):
         return True
 
-    plan, _ = build_plan(callback, None)
-    for provider in get_providers(plan):
+    for provider in get_providers(callback):
         if is_provider_using_response(provider):
             return True
 
     return False
 
 
-def get_providers(plan: andi.Plan):
-    for obj, _ in plan:
-        provider = providers.get(obj)  # type: ignore
-        if not provider:
-            continue
+def get_providers(callback):
```

Comment:
That's a good question! 

As you've noted, the previous approach used `build_plan`.  We refactored this function and it now depends on an external dependency list.  We don't have this list when we call `is_response_going_to_be_used`, which is the one using `get_providers`. 

The Injection Middleware calls this function before the download happens. This way we are able to skip it if the response is not going to be used.

Reply:
I see. Still, I think `andi.plan` can be used, something like:

```py
def get_providers(callback):
  def providers_also_injectable(cls):
     return is_injectable(cls) or cls in providers.keys()

  plan = andi.plan(callback, is_injectable=providers_also_injectable, externally_provided={Request, Response, Settings})
  return list(dict(plan).keys() & providers.keys())
```

One reason to use this approach is that all the logic with `Union` is properly managed there. For example, what to do when you have two providers in a union (e.g. `data: Union[AutoExtractResponseData, ResponseData]`).  Probably the current code of `get_providers` can be change to solve properly this case, but better to keep all this kind of code in andi internals.

Reply:
Well, even more. This piece of code https://github.com/victor-torres/scrapy-po/blob/provider-settings/scrapy_poet/utils.py#L124-L126

Can be replaced by:
```py
  def providers_also_injectable(cls):
     return is_injectable(cls) or cls in providers.keys()

  plan = andi.plan(callback, is_injectable=providers_also_injectable, externally_provided={Request, Response, Settings})
  if Request in dict(plan):
      return True
```

And then `get_providers` and `is_provider_using_response` could be removed. 

Well, and probably encapsulating this code in a function, of course.

Reply:
I was not able to make it work using your suggestion. 

```python
(Pdb) andi.plan(callback, is_injectable=lambda cls: cls in providers.keys() or is_injectable(cls), externally_provided={Response})
*** andi.errors.NonProvidableError: Not possible to generate a plan for argument 'page' in '<bound method callback_for.<locals>.parse of <MySpider 'my_spider' at 0x10ae66128>>'. Causes:
    0. Any of [<class 'str'>] types are required for argument 'url' in '<class 'web_poet.page_inputs.ResponseData'>.__init__()' but none of them is injectable or externally providable
    1. Any of [<class 'str'>] types are required for argument 'html' in '<class 'web_poet.page_inputs.ResponseData'>.__init__()' but none of them is injectable or externally providable
```

Reply:
In any case, andi's inspect function takes care of Union argument types.
See here: https://github.com/scrapinghub/andi/blob/master/andi/andi.py#L46

Reply:
> I was not able to make it work using your suggestion.

Ahh, Ok. I see the problem. Please add `ResponseData` to `externally_provided` list and try again. This should fix this problem. 

> In any case, andi's inspect function takes care of Union argument types.

Yes, but there is some logic about what to do with Unions which is in `andi.plan` here: https://github.com/scrapinghub/andi/blob/2dba95e05dd603aed0d203602e688883fc9ce1f2/andi/andi.py#L299-L319. 

For example, imagine you have two providers with a `Union` - `data: Union[AutoExtractResponseData, ResponseData]`. 
* `andi.plan` will return only one of them in the plan:
  - `AutoExtractResponseData` if it is available (if it is in providers list, or in other words, if it is in `externally_provided`)
  - `ResponseData` if it isn't
* `get_providers` will always return both with the current implementation. This will force  `Response` to be considered as required even if it is not (`AutoExtractResponseData` does not require `Response`)

The later can be fixed in `get_providers` but I think is better if we keep this kind of logic internally into `andi.plan`

Reply:
> Please add ResponseData to externally_provided list and try again. This should fix this problem.

Yes, it should. But how are we supposed to discover all those unmet dependencies during runtime?

> For example, imagine you have two providers with a Union - data: Union[AutoExtractResponseData, ResponseData].

I don't think that makes sense because a provider is registered for a single class. Maybe you're trying to refer to page objects?

> `get_providers` will always return both with the current implementation.

Yes , that's true. But so far, I've not seen a single example making use of Union. It seems like we're selecting page objects (and therefore providers) using another method. Maybe @kmike could give a word here. 

> This will force Response to be considered as required even if it is not (AutoExtractResponseData does not require Response)

Again, providers need one – and only one – page input class. I don't see this issue. But I might be biased tho.

Reply:
> But how are we supposed to discover all those unmet dependencies during runtime?

I don't get that. Could you put an example of unmet dependency?

Regarding the `Union` usage, is an original idea from @kmike so probably he can explain better. But the idea to assign `Union` an "optional" semantic. So let's say you need HTML, and you prefer it to come from the browser DOM, but you also are able to survive if it comes from direct download. You express it with the annotation `Union[SplashResponseData, ResponseData]`. In some environments, Splash will be available and in some others, it won't, but the page object will still work in both cases.

> Again, providers need one – and only one – page input class.

Yes, but this is not broken with the `Union`. The `Union` here means that one or the other will be supplied. But only one will be supplied at the end. So there are two possible providers, and only one will be invoked at the end.

Reply:
Already addressed. Thank you @victor-torres ! Resolving.